zkevm-circuits/src/blake2b/circuit/src/lib.rs

```diff
@@ -0,0 +1,1855 @@
+        // of height 8. The data of a FinalGate class gate, which are stored in xor column 
+        // triplet chunks, do not affect the gate's height (a hint: pairs.len() <= xors.len(), 
+        // so 20.0 / pairs.len() > 16.0 / xors.len())
+        let tail_height = 8 * (20.0 / pairs.len() as f64).ceil() as usize;
+        let unusable_rows = meta.blinding_factors() + 1;
+
+        let initial_input = InitialInput { r: left[0], h: hi, m: mi, t, f };
+        let rlc_table = RLCTable { allowed, selector: final_selector, rlc: second, challenge, _marker: PhantomData };
+
+        Self::Config { rlc_table, circuit_preparator, initial_gate, final_gate, round_gate, initial_input, tail_height, unusable_rows }
+    }
+
+    fn synthesize(&self, config: Self::Config, mut layouter: impl Layouter<F>) -> Result<(), Error> {
+        let initial = &config.initial_input;
+        let challenge = layouter.get_challenge(config.rlc_table.challenge);
+        
+        layouter.assign_region(|| "BLAKE2b compression function computation",
+            |mut region| {
+                config.circuit_preparator.assign(&mut region, self.k, R, R - 1, config.unusable_rows)?;
+                let mut row = R;            
+
+                for input in &self.inputs {              
+                    let mut left = F::from(input.r as u64);
+
+                    initial.r.assign(&mut region, row, Value::known(left))?;
+                    initial.t[0].assign(&mut region, row, input.t as u64)?;
+                    initial.t[1].assign(&mut region, row, (input.t >> 64) as u64)?;
+                    initial.f.assign(&mut region, row, input.f as u8)?;
+
+                    let pairs = input.h.iter().chain(input.m.iter()).zip(initial.h.iter().chain(initial.m.iter()));
+
+                    for (value, chunk) in pairs {
+                        chunk.assign(&mut region, row, *value)?;
+                    }
+
+                    let (mut h, mut m) = (input.h, input.m);
+                    let (mut rlc, mut v) = config.initial_gate.assign(&mut region, row, challenge, left, &h, &m, input.t, input.f)?;
+
+                    let mut round = 0;
+                    for _ in 0..(input.r as usize) {
+                        config.round_gate.assign(&mut region, row, rlc, &mut round, &mut left, &mut h, &mut m, &mut v)?;   
+                        row += R;
+                    }
+
+                    (h, rlc) = config.final_gate.assign(&mut region, row, challenge, rlc, &h, &v)?;
+                    // A gate of the InitialGate class is to be created next, and some of its  
+                    // data are to be stored in the 8 sequential rows before the assignment row 
+                    row += config.tail_height + 8;
+                    
+                    // Checking the correctness of the computation
+                    let (h_ex, rlc_ex) = RLCTable::compress_with_rlc(challenge, &input.h, &input.m, input.t, input.f, input.r);
```

Comment:
This correctness check should be moved elsewhere (a test case maybe?) as a malicious prover will just trim this snippet out, and it will cause overhead to honest provers. Alternatively, we can shield this block with `#[cfg(test)]`

Reply:
Is it possible to apply #[cfg(test)] not to some function, i.e. 
// Checking the correctness of the computation
"#[cfg(test)] { 
    let (h_ex, rlc_ex) = RLCTable::compress_with_rlc(challenge, &input.h, &input.m, input.t, input.f, input.r);
    let correctness = (h == h_ex) && (format!("{:?}", rlc) == format!("{:?}", rlc_ex));
    assert!(correctness, "Processing of a BLAKE2b compression function input was incorrect! This input is {:?}", input);
}"?